In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# Importing necessary Libraries for reading data

In [ ]:
import pandas as pd
import numpy as np

# Reading the dataset

In [ ]:
#reading file from drive
df1 = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Data/DL029.csv')
df1.head()

,From Date,To Date,PM2.5 (ug/m3),PM10 (ug/m3),NO (ug/m3),NO2 (ug/m3),NOx (ppb),NH3 (ug/m3),SO2 (ug/m3),CO (mg/m3),...,Benzene (ug/m3),Toluene (ug/m3),Temp (degree C),RH (%),WS (m/s),WD (degree),SR (W/mt2),BP (mmHg),AT (degree C),RF (mm)
0,2018-03-01 00:00:00,2018-03-01 01:00:00,111.67,276.18,308.18,110.22,461.12,39.13,20.41,3.70,...,0.30,3.55,NaN,88.42,0.40,200.00,2.00,989.83,18.93,NaN
1,2018-03-01 01:00:00,2018-03-01 02:00:00,123.00,303.50,204.67,94.27,37.40,29.57,10.37,3.59,...,2.06,2.52,NaN,92.00,0.48,213.83,2.00,989.08,18.28,NaN
2,2018-03-01 02:00:00,2018-03-01 03:00:00,117.58,292.25,232.13,76.03,397.35,18.57,10.22,2.47,...,0.50,4.67,NaN,93.33,0.53,217.67,2.00,989.00,17.60,NaN
3,2018-03-01 03:00:00,2018-03-01 04:00:00,111.50,261.67,166.67,73.87,302.94,37.66,16.71,2.41,...,0.20,4.32,NaN,93.75,0.48,213.25,1.92,988.25,17.13,NaN
4,2018-03-01 04:00:00,2018-03-01 05:00:00,110.08,260.83,161.93,62.19,285.57,39.52,13.46,2.26,...,0.81,3.32,NaN,96.83,0.38,208.67,2.00,988.00,16.51,NaN


In [ ]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44568 entries, 0 to 44567
Data columns (total 21 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   From Date        44568 non-null  object 
 1   To Date          44568 non-null  object 
 2   PM2.5 (ug/m3)    43892 non-null  float64
 3   PM10 (ug/m3)     43516 non-null  float64
 4   NO (ug/m3)       42857 non-null  float64
 5   NO2 (ug/m3)      43668 non-null  float64
 6   NOx (ppb)        42908 non-null  float64
 7   NH3 (ug/m3)      43420 non-null  float64
 8   SO2 (ug/m3)      43890 non-null  float64
 9   CO (mg/m3)       42448 non-null  float64
 10  Ozone (ug/m3)    43382 non-null  float64
 11  Benzene (ug/m3)  42635 non-null  float64
 12  Toluene (ug/m3)  42631 non-null  float64
 13  Temp (degree C)  0 non-null      float64
 14  RH (%)           44178 non-null  float64
 15  WS (m/s)         43884 non-null  float64
 16  WD (degree)      43869 non-null  float64
 17  SR (W/mt2)  

# Removing "To Date" column as "From Date" column contains all the necessary timestamp information and we do not need a date range, then dropping the "To Date" column and also removing 'Temp (degree C)'  as it is having 100% missing values

In [ ]:
# dropping column  to date
df1.drop('To Date', axis=1, inplace=True)

In [ ]:
# dropping Temp (degree C) column as the column is blank
df1.drop('Temp (degree C)', axis=1, inplace=True)

In [ ]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44568 entries, 0 to 44567
Data columns (total 19 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   From Date        44568 non-null  object 
 1   PM2.5 (ug/m3)    43892 non-null  float64
 2   PM10 (ug/m3)     43516 non-null  float64
 3   NO (ug/m3)       42857 non-null  float64
 4   NO2 (ug/m3)      43668 non-null  float64
 5   NOx (ppb)        42908 non-null  float64
 6   NH3 (ug/m3)      43420 non-null  float64
 7   SO2 (ug/m3)      43890 non-null  float64
 8   CO (mg/m3)       42448 non-null  float64
 9   Ozone (ug/m3)    43382 non-null  float64
 10  Benzene (ug/m3)  42635 non-null  float64
 11  Toluene (ug/m3)  42631 non-null  float64
 12  RH (%)           44178 non-null  float64
 13  WS (m/s)         43884 non-null  float64
 14  WD (degree)      43869 non-null  float64
 15  SR (W/mt2)       44090 non-null  float64
 16  BP (mmHg)        44063 non-null  float64
 17  AT (degree C

In [ ]:
#checking the % of null values in each column
(df1.isnull().sum()/df1.shape[0])*100

,0
From Date,0.000000
PM2.5 (ug/m3),1.516783
PM10 (ug/m3),2.360438
NO (ug/m3),3.839077
NO2 (ug/m3),2.019386
NOx (ppb),3.724645
NH3 (ug/m3),2.575839
SO2 (ug/m3),1.521271
CO (mg/m3),4.756776
Ozone (ug/m3),2.661102


# Feature Engineering (converting the date column to datetime format, resampling the data to daily level as the data initially was on horuly basis)

In [ ]:
# Ensure "from_date" is in datetime format
df1['From Date'] = pd.to_datetime(df1['From Date'])

# Set "from_date" as index
df1.set_index('From Date', inplace=True)

# Resample to daily level and compute mean for all columns
df1_daily = df1.resample('D').mean()

# Reset index to keep "from_date" as a column
df1_daily.reset_index(inplace=True)

# Display first few rows
print(df1_daily.head())

# Save the aggregated data
df1_daily.to_csv('daily_aggregated_data.csv', index=False)

   From Date  PM2.5 (ug/m3)  PM10 (ug/m3)  NO (ug/m3)  NO2 (ug/m3)  \
0 2018-03-01     104.800417    272.298750  124.422500    77.209167   
1 2018-03-02     111.610417    266.207500  124.567500    81.459048   
2 2018-03-03      91.795833    252.882917   61.649583    68.165417   
3 2018-03-04      76.919583    224.144583   41.310000    57.193333   
4 2018-03-05      62.364167    177.976667   43.260833    53.843333   

    NOx (ppb)  NH3 (ug/m3)  SO2 (ug/m3)  CO (mg/m3)  Ozone (ug/m3)  \
0  200.188696    38.403333    21.199583    2.065833      28.311667   
1  207.279565    37.088333    16.303750    2.280417      35.047500   
2  132.278750    35.863333    22.553750    1.454583      30.316250   
3   96.689167    33.287083    20.745833    1.007083      29.532083   
4  108.988750    30.742917    19.906250    0.977500      34.483333   

   Benzene (ug/m3)  Toluene (ug/m3)     RH (%)  WS (m/s)  WD (degree)  \
0         0.408333         2.276667  64.524167  0.632500   209.910000   
1         0.

In [ ]:
#chekcing how much % of null values are there in the aggregated data
df1_daily.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1857 entries, 0 to 1856
Data columns (total 19 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   From Date        1857 non-null   datetime64[ns]
 1   PM2.5 (ug/m3)    1854 non-null   float64       
 2   PM10 (ug/m3)     1847 non-null   float64       
 3   NO (ug/m3)       1853 non-null   float64       
 4   NO2 (ug/m3)      1853 non-null   float64       
 5   NOx (ppb)        1853 non-null   float64       
 6   NH3 (ug/m3)      1847 non-null   float64       
 7   SO2 (ug/m3)      1853 non-null   float64       
 8   CO (mg/m3)       1846 non-null   float64       
 9   Ozone (ug/m3)    1853 non-null   float64       
 10  Benzene (ug/m3)  1829 non-null   float64       
 11  Toluene (ug/m3)  1829 non-null   float64       
 12  RH (%)           1854 non-null   float64       
 13  WS (m/s)         1849 non-null   float64       
 14  WD (degree)      1849 non-null   float64

In [ ]:
df.fillna(df.mean(), inplace=True)

In [ ]:
# checking the missing rows
df4 = df1_daily[df1_daily['PM2.5 (ug/m3)'].isna()]
df4

,From Date,PM2.5 (ug/m3),PM10 (ug/m3),NO (ug/m3),NO2 (ug/m3),NOx (ppb),NH3 (ug/m3),SO2 (ug/m3),CO (mg/m3),Ozone (ug/m3),Benzene (ug/m3),Toluene (ug/m3),RH (%),WS (m/s),WD (degree),SR (W/mt2),BP (mmHg),AT (degree C),RF (mm)
763,2020-04-02,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
764,2020-04-03,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1143,2021-04-17,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


# Removing the above 3 rows where there are no values in all the columns

In [ ]:
df5 = df1_daily[~(df1_daily['PM2.5 (ug/m3)'].isna())]
df5.shape

(1854, 19)

In [ ]:
#checking the % of null values in each column
(df5.isnull().sum()/df5.shape[0])*100

,0
From Date,0.000000
PM2.5 (ug/m3),0.000000
PM10 (ug/m3),0.377562
NO (ug/m3),0.053937
NO2 (ug/m3),0.053937
NOx (ppb),0.053937
NH3 (ug/m3),0.377562
SO2 (ug/m3),0.053937
CO (mg/m3),0.431499
Ozone (ug/m3),0.053937


# SInce there are very less values in the columns so we will impute the remaining missing values with the mean

In [ ]:
df5.fillna(df5.mean(), inplace=True)

<ipython-input-24-9e8070323744>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df5.fillna(df5.mean(), inplace=True)


In [ ]:
#checking again the number of missing values and we can see that there are no missing values as the count of rows is 1854 for all the columns
df5.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1854 entries, 0 to 1856
Data columns (total 19 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   From Date        1854 non-null   datetime64[ns]
 1   PM2.5 (ug/m3)    1854 non-null   float64       
 2   PM10 (ug/m3)     1854 non-null   float64       
 3   NO (ug/m3)       1854 non-null   float64       
 4   NO2 (ug/m3)      1854 non-null   float64       
 5   NOx (ppb)        1854 non-null   float64       
 6   NH3 (ug/m3)      1854 non-null   float64       
 7   SO2 (ug/m3)      1854 non-null   float64       
 8   CO (mg/m3)       1854 non-null   float64       
 9   Ozone (ug/m3)    1854 non-null   float64       
 10  Benzene (ug/m3)  1854 non-null   float64       
 11  Toluene (ug/m3)  1854 non-null   float64       
 12  RH (%)           1854 non-null   float64       
 13  WS (m/s)         1854 non-null   float64       
 14  WD (degree)      1854 non-null   float64     

# Extarcting most important features from the dataset to avoid multicollinearity with the target variable.
Choosing Random Forest for feature selection because:

- Handles Non-Linearity & Interactions: PM2.5 levels depend on complex interactions (e.g., humidity + temperature).

- Handles Missing Data Well: Unlike regression-based methods, it can work with missing values and doesn't assume linearity.

- Ranks Features Automatically: We can directly extract importance scores without extra computations.

- Works Well for High-Dimensional Data:If we have many explanatory variables, Random Forest helps reduce dimensionality.

In [ ]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split

# Define target (dependent variable) and features (independent variables)
X = df5.drop(columns=["PM2.5 (ug/m3)", "From Date"])  # Drop PM2.5 & Date
y = df5["PM2.5 (ug/m3)"]

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train a Random Forest model
rf = RandomForestRegressor(n_estimators=100, random_state=42)
rf.fit(X_train, y_train)

# Get feature importance scores
feature_importance = pd.DataFrame({"Feature": X.columns, "Importance": rf.feature_importances_})
feature_importance = feature_importance.sort_values(by="Importance", ascending=False)

# Display the most important features
print(feature_importance)

# Optional: Save feature importance to a CSV file
feature_importance.to_csv("feature_importance.csv", index=False)

feature_importance.to_csv('/content/drive/MyDrive/Colab Notebooks/Data/feature_importance.csv', index=False)

            Feature  Importance
0      PM10 (ug/m3)    0.777483
13       SR (W/mt2)    0.083023
15    AT (degree C)    0.036175
10           RH (%)    0.022464
4       NH3 (ug/m3)    0.021258
14        BP (mmHg)    0.010624
2       NO2 (ug/m3)    0.008560
9   Toluene (ug/m3)    0.006166
6        CO (mg/m3)    0.006142
3         NOx (ppb)    0.004868
11         WS (m/s)    0.004864
8   Benzene (ug/m3)    0.004828
12      WD (degree)    0.004189
7     Ozone (ug/m3)    0.003119
5       SO2 (ug/m3)    0.002944
1        NO (ug/m3)    0.002809
16          RF (mm)    0.000486


# In the above list, PM10 (ug/m3) has the highest importance value at 0.777483, meaning it has the most significant impact on predicting PM2.5 levels. This feature should definitely be prioritized in your analysis.

Features with much lower importance (e.g., RF (mm) with an importance of 0.000486) have minimal impact on the model’s prediction, so they can be considered less important and potentially excluded to simplify the model.

Based on the importance values, these features should be prioritized:

PM10 (ug/m3) – 0.777483 (most important)
SR (W/mt2) – 0.083023
AT (degree C) – 0.036175

Other features like RH(%), NO2 (ug/m3), Toluene (ug/m3), CO (mg/m3), etc., have significantly lower importance values, so they should be considered less critical for the model’s predictive accuracy.

In summary, PM10, SR, and AT should be the primary variables, as they have the highest influence on the PM2.5 prediction.

# Downloading the final dataset after cleaning (This dataset will be used as an input to SAS time series forecasting model)

In [ ]:
df6 = df5[['From Date','PM2.5 (ug/m3)','PM10 (ug/m3)','SR (W/mt2)','AT (degree C)']]

In [ ]:
df6.to_csv('/content/drive/MyDrive/Colab Notebooks/Data/daily_aggregated_data3.csv', index=False)

# END